### ISSUE TITLE
H2 streams don't send data fairly

### ISSUE THREAD

**Issue by maskit** [apache/trafficserver#2710]
```
$ nghttp -nv -s -m 3 https://localhost:8443/bigfile
```

```
[  0.013] send HEADERS frame <length=45, flags=0x25, stream_id=13>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /bigfile
          :scheme: https
          :authority: localhost:8443
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.26.0
[  0.013] send HEADERS frame <length=19, flags=0x25, stream_id=15>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /bigfile
          :scheme: https
          :authority: localhost:8443
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.26.0
[  0.013] send HEADERS frame <length=19, flags=0x25, stream_id=17>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /bigfile
          :scheme: https
          :authority: localhost:8443
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.26.0
```

According to HEADERS frames, the three streams are equal. However, the client output was below.
```
Some requests were not processed. total=3, processed=2
***** Statistics *****

Request timing:
  responseEnd: the  time  when  last  byte of  response  was  received
               relative to connectEnd
 requestStart: the time  just before  first byte  of request  was sent
               relative  to connectEnd.   If  '*' is  shown, this  was
               pushed by server.
      process: responseEnd - requestStart
         code: HTTP status code
         size: number  of  bytes  received as  response  body  without
               inflation.
          URI: request URI

see http://www.w3.org/TR/resource-timing/#processing-model

sorted by 'complete'

id  responseEnd requestStart  process code size request path
 13    +143.28s       +376us  143.28s  200   2G /bigfile
 15    +172.60s       +422us  172.60s  200   2G /bigfile
```

It seems like ATS didn't send any data on stream 17.
```
[Oct 25 09:36:33.290] Server {0xb0016000} DEBUG: <Http2Stream.cc:319 (do_io_close)> (http2_stream) do_io_close stream 17
[Oct 25 09:36:33.290] Server {0xb0016000} DEBUG: <Http2Stream.cc:654 (destroy)> (http2_stream) Destroy stream 17, sent 0 bytes
[Oct 25 09:37:20.321] Server {0xb0016000} DEBUG: <Http2Stream.cc:266 (change_state)> (http2_stream) Http2StreamState::HTTP2_STREAM_STATE_CLOSED
[Oct 25 09:37:20.321] Server {0xb0016000} DEBUG: <Http2Stream.cc:373 (initiating_close)> (http2_stream) initiating_close stream 13
[Oct 25 09:37:20.321] Server {0xb0016000} DEBUG: <Http2Stream.cc:401 (initiating_close)> (http2_stream) handle write from destroy stream=13 event=103
[Oct 25 09:37:20.322] Server {0xb0016000} DEBUG: <Http2Stream.cc:654 (destroy)> (http2_stream) Destroy stream 13, sent 2147483648 bytes
[Oct 25 09:37:36.490] Server {0xb0016000} DEBUG: <Http2Stream.cc:266 (change_state)> (http2_stream) Http2StreamState::HTTP2_STREAM_STATE_CLOSED
[Oct 25 09:37:36.490] Server {0xb0016000} DEBUG: <Http2Stream.cc:373 (initiating_close)> (http2_stream) initiating_close stream 15
[Oct 25 09:37:36.490] Server {0xb0016000} DEBUG: <Http2Stream.cc:401 (initiating_close)> (http2_stream) handle write from destroy stream=15 event=103
[Oct 25 09:37:36.490] Server {0xb0016000} DEBUG: <Http2Stream.cc:654 (destroy)> (http2_stream) Destroy stream 15, sent 2147483648 bytes
```

**Comment by maskit** [comment]
proxy.config.http2.stream_priority_enabled was 0

**Comment by masaori335** [comment]
I tried with `--no-dep`, and also got similar result.
```
$ nghttp -vn -s -m3 --no-dep https://localhost:4443/1g.txt
[snip]
Some requests were not processed. total=3, processed=2
***** Statistics *****

Request timing:
  responseEnd: the  time  when  last  byte of  response  was  received
               relative to connectEnd
 requestStart: the time  just before  first byte  of request  was sent
               relative  to connectEnd.   If  '*' is  shown, this  was
               pushed by server.
      process: responseEnd - requestStart
         code: HTTP status code
         size: number  of  bytes  received as  response  body  without
               inflation.
          URI: request URI

see http://www.w3.org/TR/resource-timing/#processing-model

sorted by 'complete'

id  responseEnd requestStart  process code size request path
  1    +155.81s       +208us  155.81s  200   1G /1g.txt
  3    +201.40s       +244us  201.40s  200   1G /1g.txt
```

**Comment by masaori335** [comment]
When I disable `transaction_no_activity_timeout_in`, I got all of them.
```
CONFIG proxy.config.http.transaction_no_activity_timeout_in INT 0
```

```
***** Statistics *****

Request timing:
  responseEnd: the  time  when  last  byte of  response  was  received
               relative to connectEnd
 requestStart: the time  just before  first byte  of request  was sent
               relative  to connectEnd.   If  '*' is  shown, this  was
               pushed by server.
      process: responseEnd - requestStart
         code: HTTP status code
         size: number  of  bytes  received as  response  body  without
               inflation.
          URI: request URI

see http://www.w3.org/TR/resource-timing/#processing-model

sorted by 'complete'

id  responseEnd requestStart  process code size request path
 13    +131.13s       +175us  131.13s  200   1G /1g.txt
 15    +166.18s       +299us  166.17s  200   1G /1g.txt
 17    +204.03s       +329us  204.03s  200   1G /1g.txt
```

**Comment by maskit** [comment]
It seems that the first two streams consume all window and the rest streams can't send data at all.

**Comment by maskit** [comment]
This affects both the master and 7.1.x. Probably 6.2.x too.

Workarounds are:
- Set proxy.config.http.transaction_no_activity_timeout_in to 0 or long enough
or
- Set proxy.config.http2.stream_priority_enabled to 1 (This may decrease performance)
or
- Serve only small files

**Comment by scw00** [comment]
It seems we use the `stream id` as the `node->point` and update the point with the length of date we send. So the lower stream id will get more chance to send data.

```
Node(uint32_t i, uint32_t w, uint32_t p, Node *n, void *t = nullptr) : id(i), weight(w), point(p), t(t), parent(n)

 node = new Node(id, weight, id, parent, t);
```

```
void
Http2ConnectionState::send_data_frames_depends_on_priority()
{
  Http2DependencyTree::Node *node = dependency_tree->top();

  // No node to send or no connection level window left
  if (node == nullptr || client_rwnd <= 0) {
    return;
  }

  Http2Stream *stream = static_cast<Http2Stream *>(node->t);
  ink_release_assert(stream != nullptr);
  DebugHttp2Stream(ua_session, stream->get_id(), "top node, point=%d", node->point);

  size_t len                       = 0;
  Http2SendADataFrameResult result = send_a_data_frame(stream, len);

  switch (result) {
  case HTTP2_SEND_A_DATA_FRAME_NO_ERROR: {
    // No response body to send
    if (len == 0 && !stream->is_body_done()) {
      dependency_tree->deactivate(node, len);
    } else {
      dependency_tree->update(node, len);
    }
    break;
  }
```

May relate to this issue !

**Comment by masaori335** [comment]
@scw00 This happens **without** priority tree. If it enabled, streams are handled fairly.

**Comment by maskit** [comment]
@scw00 You can see what's happening with this patch:

```diff
diff --git a/proxy/http2/Http2ConnectionState.cc b/proxy/http2/Http2ConnectionState.cc
index 225a1c099..76791a4f0 100644
--- a/proxy/http2/Http2ConnectionState.cc
+++ b/proxy/http2/Http2ConnectionState.cc
@@ -1312,7 +1312,11 @@ Http2ConnectionState::send_data_frames(Http2Stream *stream)
       this->delete_stream(stream);
       break;
     } else if (result == HTTP2_SEND_A_DATA_FRAME_NO_ERROR) {
+      DebugSsn(this->ua_session, "http2_stream", "-------------------  NO ERROR%d", stream->get_id());
       continue;
+    } else if (result == HTTP2_SEND_A_DATA_FRAME_NO_WINDOW) {
+      DebugSsn(this->ua_session, "http2_stream", "-------------------  NO WINDOW %d", stream->get_id());
+      break;
     } else {
       break;
     }
```

**Comment by scw00** [comment]
that's great ! I'll do the test .

**Comment by masaori335** [comment]
When ATS received connection level WINDOW_UPDATE frame, [`Http2ConnectionState::restart_streams()`](https://github.com/apache/trafficserver/blob/46a7fb850e1db377bb1714707baafa442df60c6d/proxy/http2/Http2ConnectionState.cc#L1068) is called. It restart sending response body from the head of `stream_list`. This is why first 2 streams always consume all window.

**Comment by maskit** [comment]
I have a fix for this but need to fix #2737 first.
